src/script/sign.h

```diff
@@ -73,6 +76,538 @@ struct SignatureData {
+    PSBTOutput() {}
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        // Write the redeem script
+        if (!redeem_script.empty()) {
+            SerializeToVector(s, PSBT_OUT_REDEEMSCRIPT);
+            s << redeem_script;
+        }
+
+        // Write the witness script
+        if (!witness_script.empty()) {
+            SerializeToVector(s, PSBT_OUT_WITNESSSCRIPT);
+            s << witness_script;
+        }
+
+        // Write any hd keypaths
+        SerializeHDKeypaths(s, hd_keypaths, PSBT_OUT_BIP32_DERIVATION);
+
+        // Write unknown things
+        for (auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        s << PSBT_SEPARATOR;
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        while(!s.empty()) {
+            // Read
+            std::vector<unsigned char> key;
+            s >> key;
+
+            // the key is empty if that was actually a separator byte
+            // This is a special case for key lengths 0 as those are not allowed (except for separator)
+            if (key.empty()) return;
+
+            // First byte of key is the type
+            unsigned char type = key[0];
+
+            // Do stuff based on type
+            switch(type) {
+                case PSBT_OUT_REDEEMSCRIPT:
+                {
+                    if (!redeem_script.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, output redeemScript already provided");
+                    }
+                    s >> redeem_script;
+                    break;
+                }
+                case PSBT_OUT_WITNESSSCRIPT:
+                {
+                    if (!witness_script.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, output witnessScript already provided");
+                    }
+                    s >> witness_script;
+                    break;
+                }
+                case PSBT_OUT_BIP32_DERIVATION:
+                {
+                    DeserializeHDKeypaths(s, key, hd_keypaths);
+                    break;
+                }
+                // Unknown stuff
+                default: {
+                    if (unknown.count(key) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, key for unknown value already provided");
+                    }
+                    // Read in the value
+                    std::vector<unsigned char> val_bytes;
+                    s >> val_bytes;
+                    unknown.emplace(std::move(key), std::move(val_bytes));
+                    break;
+                }
+            }
+        }
+    }
+
+    template <typename Stream>
+    PSBTOutput(deserialize_type, Stream& s) {
+        Unserialize(s);
+    }
+};
+
+/** A version of CTransaction with the PSBT format*/
+struct PartiallySignedTransaction
+{
+    CMutableTransaction tx;
```

Comment:
Is this were a `std::unique_ptr<CMutableTransaction>` (or `boost::optional<CMutableTransaction>`), you could avoid the need for implementing `IsNull` or converting to `CTransaction` to decide whether it is set or not.

Reply:
Made into a `boost::optional`